beetsplug/aura.py

```diff
@@ -0,0 +1,978 @@
+        if image_path is None:
+            return None
+
+        attributes = {
+            "role": "cover",
+            "mimetype": guess_type(image_path)[0],
+            "size": getsize(image_path),
+        }
+        try:
+            from PIL import Image
+        except ImportError:
+            pass
+        else:
+            im = Image.open(image_path)
+            attributes["width"] = im.width
+            attributes["height"] = im.height
+
+        relationships = {}
+        # Split id into [parent_type, parent_id, filename]
+        id_split = image_id.split("-")
+        relationships[id_split[0] + "s"] = {
+            "data": [{"type": id_split[0], "id": id_split[1]}]
+        }
+
+        return {
+            "id": image_id,
+            "type": "image",
+            # Remove attributes that are None, 0, "", etc.
+            "attributes": {k: v for k, v in attributes.items() if v},
+            "relationships": relationships,
+        }
+
+    def single_resource(self, image_id):
+        """Get info for the requested image and build a document.
+
+        Args:
+            image_id: A string in the form
+                "<parent_type>-<parent_id>-<img_filename>".
+        """
+        image_resource = self.resource_object(image_id)
+        if image_resource is None:
+            return self.error(
+                "404 Not Found",
+                "No image with the requested id.",
+                ("There is no image with an id of {} in the library.").format(
+                    image_id
+                ),
+            )
+        return self.single_resource_document(image_resource)
+
+
+# Initialise flask blueprint
+aura_bp = Blueprint("aura_bp", __name__)
+
+
+@aura_bp.route("/server")
+def server_info():
+    """Respond with info about the server."""
+    return {"data": {"type": "server", "id": "0", "attributes": SERVER_INFO}}
+
+
+# Track endpoints
+
+
+@aura_bp.route("/tracks")
+def all_tracks():
+    """Respond with a list of all tracks and related information."""
+    doc = TrackDocument()
+    return doc.all_resources()
+
+
+@aura_bp.route("/tracks/<int:track_id>")
+def single_track(track_id):
+    """Respond with info about the specified track.
+
+    Args:
+        track_id: The id of the track provided in the URL (integer).
+    """
+    doc = TrackDocument()
+    return doc.single_resource(track_id)
+
+
+@aura_bp.route("/tracks/<int:track_id>/audio")
+def audio_file(track_id):
+    """Supply an audio file for the specified track.
+
+    Args:
+        track_id: The id of the track provided in the URL (integer).
+    """
+    track = current_app.config["lib"].get_item(track_id)
+    if track is None:
+        return AURADocument.error(
+            "404 Not Found",
+            "No track with the requested id.",
+            ("There is no track with an id of {} in the library.").format(
+                track_id
+            ),
+        )
+
+    path = displayable_path(track.destination())
```

Comment:
Flask's [send_file](https://flask.palletsprojects.com/en/1.1.x/api/?highlight=send_file#flask.send_file) breaks if it's given a bytes object as a file name. I think it does some parsing assuming filename is a string.

But you're right, `displayable_path` shouldn't be necessary because `Item.destination()` returns unicode. However for me it still returns bytes (haven't figured out what's going on there though).

Or I could change it to `displayable_path(track.path)`, which works fine.

Reply:
Got it; makes sense to me! I think the right thing here is `py3_path(track.path)`. There are two parts to this:

- The reason to use `track.path` is because that's where the file _is actually located right now_, as opposed to where it _would be_ located if the user did a `beet move` to consolidate their library. Since I think we want the actual file, that's the right thing here.
- While both `displayable_path` and `py3_path` convert paths from `bytes` to `unicode`, they are subtly different: `displayable_path` does its best to make something that will be viewable by the user, at the expense of lossiness. So "bad" characters for the encoding will just be removed entirely, yielding a wrong path that nonetheless might be useful for a human to look at. On the other hand `py3_path` is for cases like this: when a Python 3-era API only accepts surrogate-escaped Unicode instead of both text and bytes (as most standard-library APIs do, for example).

Reply:
Ah, right; I didn't realise the difference between them. I'll change it to 'py3_path(track.path)`